pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayLiteralTransformFunction.java

```diff
@@ -0,0 +1,285 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import com.google.common.base.Preconditions;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.operator.ColumnContext;
+import org.apache.pinot.core.operator.blocks.ValueBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+import org.roaringbitmap.RoaringBitmap;
+
+
+/**
+ * The <code>LiteralTransformFunction</code> class is a special transform function which is a wrapper on top of a
+ * LITERAL. The data type is inferred from the literal string.
+ */
+public class ArrayLiteralTransformFunction implements TransformFunction {
+  public static final String FUNCTION_NAME = "arrayValueConstructor";
+
+  private final DataType _dataType;
+
+  private final int[] _intArrayLiteral;
+  private final long[] _longArrayLiteral;
+  private final float[] _floatArrayLiteral;
+  private final double[] _doubleArrayLiteral;
+  private final String[] _stringArrayLiteral;
+
+  // literals may be shared but values are intentionally not volatile as assignment races are benign
+  private int[][] _intArrayResult;
+  private long[][] _longArrayResult;
+  private float[][] _floatArrayResult;
+  private double[][] _doubleArrayResult;
+  private String[][] _stringArrayResult;
+
+  public ArrayLiteralTransformFunction(List<ExpressionContext> literalContexts) {
+    Preconditions.checkNotNull(literalContexts);
+    if (literalContexts.isEmpty()) {
+      _dataType = DataType.UNKNOWN;
+      _intArrayLiteral = new int[0];
+      _longArrayLiteral = new long[0];
+      _floatArrayLiteral = new float[0];
+      _doubleArrayLiteral = new double[0];
+      _stringArrayLiteral = new String[0];
+      return;
+    }
+    _dataType = literalContexts.get(0).getLiteral().getType();
```

Comment:
We don't know it is a literal yet. Preconditions check is needed

Reply:
moved the check inside this init()